pkg/assembler/backends/testing/isOccurrence.go

```diff
@@ -17,209 +17,343 @@ package testing
+	// selectedType := "conan"
+	// selectedNameSpace := "openssl.org"
+	// selectedName := "openssl"
+	// selectedVersion := "3.0.3"
+	// qualifierA := "bincrafters"
+	// qualifierB := "stable"
+	// selectedQualifiers := []*model.PackageQualifierSpec{{Key: "user", Value: &qualifierA}, {Key: "channel", Value: &qualifierB}}
+	// selectedPkgSpec := &model.PkgSpec{Type: &selectedType, Namespace: &selectedNameSpace, Name: &selectedName, Version: &selectedVersion, Qualifiers: selectedQualifiers}
+	// selectedPackage, err := client.Packages(context.TODO(), selectedPkgSpec)
+	// if err != nil {
+	// 	return err
+	// }
+	// _, err = client.registerIsOccurrence(selectedPackage[0], nil, &model.Artifact{Digest: "5a787865sd676dacb0142afa0b83029cd7befd9", Algorithm: "sha1"}, "this artifact is an occurrence of this package", "testing backend", "testing backend")
+	// if err != nil {
+	// 	return err
+	// }
+	// // "git", "github", "github.com/guacsec/guac", "tag=v0.0.1"
+	// selectedSourceType := "git"
+	// selectedSourceNameSpace := "github"
+	// selectedSourceName := "github.com/guacsec/guac"
+	// selectedTag := "v0.0.1"
+	// selectedSourceSpec := &model.SourceSpec{Type: &selectedSourceType, Namespace: &selectedSourceNameSpace, Name: &selectedSourceName, Tag: &selectedTag}
+	// //selectedSource, err := client.Sources(context.TODO(), selectedSourceSpec)
+	// _, err = client.Sources(context.TODO(), selectedSourceSpec)
+	// if err != nil {
+	// 	return err
+	// }
+	// //_, err = client.registerIsOccurrence(nil, selectedSource[0], client.artifacts[0], "this artifact is an occurrence of this source", "testing backend", "testing backend")
+	// if err != nil {
+	// 	return err
+	// }
 	return nil
 }
 
 // Ingest IsOccurrence
+func (c *demoClient) IngestOccurrence(ctx context.Context, subject model.PackageOrSourceInput, artifact model.ArtifactInputSpec, occurrence model.IsOccurrenceInputSpec) (*model.IsOccurrence, error) {
+	err := helper.ValidatePackageOrSourceInput(&subject, "IngestOccurrence")
+	if err != nil {
+		return nil, err
+	}
 
-func (c *demoClient) registerIsOccurrence(selectedPackage *model.Package, selectedSource *model.Source, artifact *model.Artifact, justification, origin, collector string) (*model.IsOccurrence, error) {
-
-	if selectedPackage != nil && selectedSource != nil {
-		return nil, fmt.Errorf("cannot specify both package and source for IsOccurrence")
+	a, err := c.artifactByKey(artifact.Algorithm, artifact.Digest)
+	if err != nil {
+		return nil, gqlerror.Errorf("IngestOccurrence :: Artifact not found")
 	}
 
-	for _, occurrence := range c.isOccurrence {
-		if reflect.DeepEqual(occurrence.Artifact, artifact) && occurrence.Justification == justification {
-			if val, ok := occurrence.Subject.(model.Package); ok {
-				if reflect.DeepEqual(val, *selectedPackage) {
-					return occurrence, nil
+	// FIXME these sections are copied from ingestPackage/Source or hasSourceAt
+	packageID := maxUint32
+	if subject.Package != nil {
```

Comment:
there is a helper function for this now so there is no longer a need to repeat the code.

Reply:
done